packages/foundry/contracts/WitchBase.sol

```diff
@@ -0,0 +1,680 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity >=0.8.13;
+
+import "@yield-protocol/utils-v2/contracts/access/AccessControl.sol";
+import "@yield-protocol/utils-v2/contracts/cast/CastU256U128.sol";
+import "@yield-protocol/utils-v2/contracts/math/WMul.sol";
+import "@yield-protocol/utils-v2/contracts/math/WDiv.sol";
+import "@yield-protocol/utils-v2/contracts/math/WDivUp.sol";
+import "./interfaces/DataTypes.sol";
+import "./interfaces/ILadle.sol";
+import "./interfaces/ICauldron.sol";
+import "./interfaces/IJoin.sol";
+
+contract WitchBase is AccessControl {
+    using WMul for uint256;
+    using WDiv for uint256;
+    using WDivUp for uint256;
+    using CastU256U128 for uint256;
+    // ==================== Errors ====================
+
+    error VaultAlreadyUnderAuction(bytes12 vaultId, address witch);
+    error VaultNotLiquidatable(bytes6 ilkId, bytes6 baseId);
+    error AuctionIsCorrect(bytes12 vaultId);
+    error AuctioneerRewardTooHigh(uint256 max, uint256 actual);
+    error WitchIsDead();
+    error CollateralLimitExceeded(uint256 current, uint256 max);
+    error NotUnderCollateralised(bytes12 vaultId);
+    error UnderCollateralised(bytes12 vaultId);
+    error VaultNotUnderAuction(bytes12 vaultId);
+    error NotEnoughBought(uint256 expected, uint256 got);
+    error JoinNotFound(bytes6 id);
+    error UnrecognisedParam(bytes32 param);
+    error LeavesDust(uint256 remainder, uint256 min);
+
+    // ==================== User events ====================
+
+    event Auctioned(
+        bytes12 indexed vaultId,
+        DataTypes.Auction auction,
+        uint256 duration,
+        uint256 initialCollateralProportion
+    );
+    event Cancelled(bytes12 indexed vaultId);
+    event Cleared(bytes12 indexed vaultId);
+    event Ended(bytes12 indexed vaultId);
+    event Bought(
+        bytes12 indexed vaultId,
+        address indexed buyer,
+        uint256 ink,
+        uint256 art
+    );
+
+    // ==================== Governance events ====================
+
+    event Point(
+        bytes32 indexed param,
+        address indexed oldValue,
+        address indexed newValue
+    );
+    event LineSet(
+        bytes6 indexed ilkId,
+        bytes6 indexed baseId,
+        uint32 duration,
+        uint64 vaultProportion,
+        uint64 collateralProportion
+    );
+    event LimitSet(bytes6 indexed ilkId, bytes6 indexed baseId, uint128 max);
+    event ProtectedSet(address indexed value, bool protected);
+    event AuctioneerRewardSet(uint256 auctioneerReward);
```

Comment:
Something I recently learned is that it's better to have errors and events in the interface, 2 main reasons
- the ABI of the interface is self-sufficient
- you can split the interface in 2 (one for events and other for methods & errors) and make the tests extend the former, so you avoid duplicating events on the tests and al de headaches that that creates